edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Amqp/CbsNode.cs

```diff
@@ -73,33 +75,47 @@ public void RegisterLink(IAmqpLink link)
             Events.LinkRegistered(link);
         }
 
-        public async Task<bool> AuthenticateAsync(string id, Option<string> modelId)
+        public async Task<bool> AuthenticateAsync(string id, Option<string> modelId, Option<string> authChain)
         {
             try
             {
-                if (this.clientCredentialsMap.TryGetValue(id, out CredentialsInfo credentialsInfo))
+                // See if we received direct credentials for the target identity
+                ClientToken credentialsInfo;
+                bool hasCredentials = this.clientTokens.TryGetValue(id, out credentialsInfo);
+
+                // Otherwise, this could be a child Edge acting OnBehalfOf of the target,
+                // in which case we would have the actor's credentials instead
+                if (!hasCredentials)
                 {
-                    // Not -ve caching isAuthenticated here.
-                    // If incorrect credentials are sent, then it authenticates every time and fails
-                    // If correct credentials are sent later, then the authentication will succeed.
-                    if (!credentialsInfo.IsAuthenticated)
+                    Option<string> actorDeviceId = AuthChainHelpers.GetActorDeviceId(authChain);
+                    hasCredentials = actorDeviceId.Map(actor => this.clientTokens.TryGetValue(actor + $"/{CoreConstants.EdgeHubModuleId}", out credentialsInfo)).GetOrElse(false);
+                }
+
+                if (hasCredentials)
+                {
+                    if (!authenticatedClients.ContainsKey(id))
```

Comment:
Do we need to lock here? I am not sure if the ConcurrentDictionary suffices..

Reply:
Read/write on a concurrent collection itself is threadsafe, right?  Since we're never modifying the stored value, the keys should be safe to read from multiple threads.

---
In reply to: [489200704](https://github.com/Azure/iotedge/pull/3552#discussion_r489200704) [](ancestors = 489200704)

Reply:
Talked offline, the concern is around re-entrancy of the method itself, added a lock.